### PR TITLE
Improve arity error messages

### PR DESCRIPTION
**Pull request by sandersn**
Add a specific errors for incorrect number of arguments or of type arguments in `resolveCall`. I'm not sure how we lived without this before.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15380,19 +15380,24 @@ namespace ts {
...
-                    "a minimum of " + (args.length - 1) :
-                    args.length.toString();
                 let min = Number.POSITIVE_INFINITY;
                 let max = Number.NEGATIVE_INFINITY;
                 for (const sig of signatures) {
                     min = Math.min(min, sig.minArgumentCount);
                     max = Math.max(max, sig.parameters.length);
                 }
-                const paramMessage = some(signatures, sig => sig.hasRestParameter) ? "at least " + min :
+                ///////////
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Just add a newline?

**Comment by sandersn** [comment]
@RyanCavanaugh @DanielRosenwasser mind taking a look? This PR is pain to keep up to date because it changes so many error numbers ...

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -1859,10 +1859,22 @@
...
-    "Expected {0} type arguments, but got {1}.": {
+    "Expected at least {0} arguments, but got {1}.": {
         "category": "Error",
         "code": 2554
     },
+    "Expected {0} arguments, but got a minimum of {1}.": {
+        "category": "Error",
+        "code": 2555
+    },
+    "Expected at least {0} arguments, but got a minimum of {1}.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: What does this mean? I can't tell from the message nor the condition in the code. Also, every error message we report this on seems to have "a minimum of 0"

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -1859,10 +1859,22 @@
         "category": "Error",
         "code": 2553
     },
-    "Expected {0} type arguments, but got {1}.": {
+    "Expected at least {0} arguments, but got {1}.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: `"Expected {0} type arguments, but got {1}."` is closer to the old message. I'd argue that the error code didn't need to change. Consider switching this with the message assigned to code 2557

File: src/compiler/diagnosticMessages.json [hunk 4]
```diff
@@ -1859,10 +1859,22 @@
...
     },
+    "Expected {0} arguments, but got a minimum of {1}.": {
+        "category": "Error",
+        "code": 2555
+    },
+    "Expected at least {0} arguments, but got a minimum of {1}.": {
+        "category": "Error",
+        "code": 2556
+    },
+    "Expected {0} type arguments, but got {1}.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: In some cases, type arguments also need "at least" some number. Is that something that would be done in a different PR?

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/checker.ts]: Why not newlines? 🦀 👽 🦑

File: src/compiler/diagnosticMessages.json (hunk 4 above)
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: "at least" indicates a rest argument, which TS doesn't have yet. I use a range, generated in the code, for optional [type] arguments, so you'd see a message like "Expected 1-3 type arguments, but got 4".

File: src/compiler/diagnosticMessages.json (hunk 2 above)
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: I used "a minimum of" for spreads to indicate that a spread adds 0 or more arguments to those already present. After reading the example you might have a better idea for the wording.

Here's an example that would give "a minimum of 1":

```ts
declare function f(first, second, ...rest): any[]
declare const car: number
declare const cdr: number[]
f(car, ...cdr) // Expected at least 2 arguments, but got a minimum of 1.
```

Note that this is legal Javascript and could succeed at runtime, but we can't prove that it will because `cdr` might be empty.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/checker.ts]: (Removed in the current version)

**Comment by sandersn** [comment]
@DanielRosenwasser I replied to your now-hidden replies, but my replies are now also hidden. (!)

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: I think the diff is wonky here. All these messages are new, so there aren't any codes to reuse.
